src/sdk/conference/channel.js

```diff
@@ -190,111 +200,159 @@ export class ConferencePeerConnectionChannel extends EventDispatcher {
-            this._pc.addTrack(track, stream.mediaStream);
-          }
+        const transceiver = this._pc.addTransceiver(
+            stream.mediaStream.getVideoTracks()[0],
+            transceiverInit);
+        transceivers.push({
+          type: 'video',
+          transceiver,
+          source: mediaOptions.video.source,
+          option: {video: options.video},
+        });
+
+        if (Utils.isFirefox()) {
+          // Firefox does not support encodings setting in addTransceiver.
+          const parameters = transceiver.sender.getParameters();
+          parameters.encodings = transceiverInit.sendEncodings;
+          await transceiver.sender.setParameters(parameters);
         }
-        offerOptions.offerToReceiveAudio = false;
-        offerOptions.offerToReceiveVideo = false;
-      }
-      return offerOptions;
-    }).then((offerOptions) => {
-      let localDesc;
-      this._pc.createOffer(offerOptions).then((desc) => {
-        if (options) {
-          desc.sdp = this._setRtpReceiverOptions(desc.sdp, options);
+      }
+    } else {
+      // Should not reach here
+      if (mediaOptions.audio &&
+          stream.mediaStream.getAudioTracks().length > 0) {
+        for (const track of stream.mediaStream.getAudioTracks()) {
+          this._pc.addTrack(track, stream.mediaStream);
         }
-        return desc;
-      }).then((desc) => {
-        localDesc = desc;
-        return this._pc.setLocalDescription(desc);
-      }).then(() => {
-        this._signaling.sendSignalingMessage('soac', {
-          id: this
-              ._internalId,
-          signaling: localDesc,
+      }
+      if (mediaOptions.video &&
+          stream.mediaStream.getVideoTracks().length > 0) {
+        for (const track of stream.mediaStream.getVideoTracks()) {
+          this._pc.addTrack(track, stream.mediaStream);
+        }
+      }
+      offerOptions.offerToReceiveAudio = false;
+      offerOptions.offerToReceiveVideo = false;
+    }
+
+    const internalId = this._createInternalId();
+    this._publishTransceivers.set(internalId, {transceivers});
+
+    // Waiting for previous SDP negotiation if needed
+    await this._chainSdpPromise();
+
+    let localDesc;
+    this._pc.createOffer(offerOptions).then((desc) => {
+      localDesc = desc;
+      return this._pc.setLocalDescription(desc);
+    }).then(() => {
+      const trackOptions = [];
+      transceivers.forEach(({type, transceiver, source}) => {
+        trackOptions.push({
+          type,
+          mid: transceiver.mid,
+          source,
         });
-      }).catch((e) => {
-        Logger.error('Failed to create offer or set SDP. Message: '
-            + e.message);
-        this._unpublish();
-        this._rejectPromise(e);
-        this._fireEndedEventOnPublicationOrSubscription();
+      });
+      return this._signaling.sendSignalingMessage('publish', {
+        media: {tracks: trackOptions},
+        attributes: stream.attributes,
+        transportId: this._id,
```

Comment:
FYI. This property is changed to an object `{type: 'webrtc' or 'quic', id: string}` in my PR for WebTransport support at server side. It may break this PR.